### PR TITLE
[ADD] Feature to add multi images for single item and multi image zoo…

### PR DESCRIPTION
**Pull request by Solufyin**
@revant, @manassolanki 
Feature to add multi images to a single item and multi-image zooming functionality.
add_multiple_images_for_single_item

item_multi_image_zooming

See the attached video for more information :

[ERPNext_product_muti_image_zooming.mp4.zip](https://github.com/frappe/erpnext/files/1677505/ERPNext_product_muti_image_zooming.mp4.zip)

### CODE REVIEW

**Comment by revant** [comment]
I think the data on `Item Multi Image` is already captured on `Website Slideshow` DocType from frappe.
screenshot from 2018-01-30 18-36-13

`Website Slideshow` Feature needs update?

**Comment by Solufyin** [comment]
Hello, @revant,
We keep both features when we configure "Item Multi Image" It will show the Item Multi Zooming and we configure item slideshow it will show the slideshow on e-commerce site.

The User can use two alternative choices:
1) If user added multiple images in "Item Multi Image" then user is able to see product multi 
zooming functionality on e-commerce site.

2) If the user added multiple images in "Slider" ( for slideshow ) then a user is able to see slideshow functionality on e-commerce site.

Thanks

**Comment by rmehta** [comment]
"Multi Item Image" is a very bad name!

How about "Alternate Images" ?

File: erpnext/stock/doctype/item/item.js [hunk 1]
```diff
@@ -119,6 +119,50 @@ frappe.ui.form.on("Item", {
...
+
+		if (frm.doc.list_all_image.length == 0)
+		{
+			frappe.msgprint(__("Images not avalilable !!"));
+		}
+		else if (frm.doc.list_all_image){
+				var image_path = frm.doc.image;
+				fields = []
+				for(var i=0;i< frm.doc.list_all_image.length;i++){
+//					fieldname = 'image'
```
Comment (rmehta) [inline erpnext/stock/doctype/item/item.js]: remove commented code

File: erpnext/stock/doctype/item/item.js [hunk 2]
```diff
@@ -119,6 +119,50 @@ frappe.ui.form.on("Item", {
...
+//					fieldname = 'title'
+//					fieldname = 'description'
+				    image_dict = {
```
Comment (rmehta) [inline erpnext/stock/doctype/item/item.js]: why this modal?

Also don't use variable names like `image_dict1` etc. This should just be one list of dicts?

File: erpnext/stock/doctype/item/item.js [hunk 3]
```diff
@@ -119,6 +119,50 @@ frappe.ui.form.on("Item", {
...
+					fields.push(image_dict1);
+					fields.push(image_dict3);
+				}
+			}
+			var d = new frappe.ui.Dialog({
+				fields: fields
+			});
+			d.show();
+			$('.control-value.like-disabled-input').css('display','block');
+			$("#myImg").css('display','block');
```
Comment (rmehta) [inline erpnext/stock/doctype/item/item.js]: no absolute links please!

File: erpnext/templates/generators/item.html [hunk 4]
```diff
@@ -11,92 +11,203 @@
...
-			<div class="col-sm-6">
-				{% if slideshow %}
-					{% include "templates/includes/slideshow.html" %}
-				{% else %}
-					{{ product_image(website_image, "product-full-image") }}
-				{% endif %}
-			</div>
-			<div class="col-sm-6" style="padding-left:20px;">
-				 <h2 itemprop="name" style="margin-top: 0px;">{{ item_name }}</h2>
+		    {% if not list_all_image %}
```
Comment (rmehta) [inline erpnext/templates/generators/item.html]: whitespace

File: erpnext/stock/doctype/item/item.json [hunk 5]
```diff
@@ -701,6 +701,97 @@
...
+   "ignore_user_permissions": 0, 
+   "ignore_xss_filter": 0, 
+   "in_filter": 0, 
+   "in_global_search": 0, 
+   "in_list_view": 0, 
+   "in_standard_filter": 0, 
+   "label": "List All Image", 
+   "length": 0, 
+   "no_copy": 0, 
+   "options": "Item Multi Image",
```
Comment (rmehta) [inline erpnext/stock/doctype/item/item.json]: Item Alternate Image

https://github.com/frappe/erpnext/wiki/Naming-Guidelines

**Review by rmehta — CHANGES_REQUESTED** [review]

**Comment by tundebabzy** [comment]
Hi @Solufyin 
I'm not sure if this is the right approach to this new feature (which I like). How does this affect POS?

**Review by tundebabzy — DISMISSED** [review]
Fix codacy review also

**Comment by hda** [comment]
@tundebabzy POS image and E-commerce image will not have conflicts as we are storing e-commerce alternate images in separate doctype. We are in discussion on this to use slideshow doctype instead of new doctype as suggested by @revant

**Comment by MichaelPinkowski** [comment]
I strongly prefer the idea of storing e-commerce images separately from the slideshow.

**Comment by rmehta** [comment]
I think we can keep images separate too.

**Comment by revant** [comment]
Note : 

I've cases where some DocType need multiple images attached to it.

- Slideshow was separate (and part of framework), 
- Any DocType can have link to Website Slideshow,
- linked Slideshow will only contain "media" / images and not any non image files.

Cases.

- Merchandise posted by a user/seller on e-commerce Portal.
- List of previous Profile photos.

I feel better file upload, and file meta management is required to manage file backend (not UI).

I could achieve this using [whitelisted api](https://github.com/agrinext/agrinext/blob/develop/agrinext/api.py#L25):
(I'm testing the whitelisted api for permissions, eventually I'll send PR)

Multiple files attached images and other files.
photo_2018-02-01_12-21-39

[Android App](https://github.com/agrinext/agrinext_mobile/tree/revant-dev) (UI) shows previous profile images

**Comment by rmehta** [comment]
I also like @revant suggestion of making this generic.

We can rename "Website Slideshow" to "Media Group" and create a new Media Group object for each Item and link it to the Item.

In the template, the workflow will be the same, either you pull it from Item or Media Group?

**Comment by Solufyin** [comment]
Hello @revant, @rmehta,
We did improvements as per your suggestion.

**Comment by rmehta** [comment]
@Solufyin thanks for the fixes. I agree with @revant  that we can reuse "Website Slideshow" and link it in the Item instead of creating a new DocType. 

Just remove the Item Alternate table and link "Website Slideshow"

Bonus if you can rename "Website Slideshow" to "Media Group" in frappe.

**Comment by revant** [comment]
@Solufyin,

> Just remove the Item Alternate table and link "Website Slideshow"

Can you do the suggested change?

We'll eventually rename `Website Slideshow` to `Media Group`. That refactor will have to be patched for old databases. Either you can attempt that on frappe/frappe, or just use `Website Slideshow` for now.

**Comment by Solufyin** [comment]
**Dear @Revant and @rmehta,**

We are Improved Code for Item Alternate Zooming functionality using common doctype website slideshow.
We can use multiple choice as per user need using common doctype "Website Slideshow Item".

**Comment by revant** [comment]
@Solufyin 

There seems to be a merge conflict with item.json

What I suggest this as a workaround:

- Note down fields added in item.json as a part of this PR
- pull latest item.json from upstream/develop
- `bench migrate`
- Add the fields noted down before (if not there in `Item` `DocType`)
- save Item DocType, commit as conflict resolved

**Comment by Solufyin** [comment]
Hello @revant ,
We resolved conflict item.json.
Thanks.

**Comment by revant** [comment]
@netchampfaris, can you help in review of CSS/HTML/JS?

@Solufyin I cleared the print statement do pull before you start making any changes. Also there seems to be merge conflict item.json again. :frowning_face:

**Comment by netchampfaris** [comment]
@Solufyin 

After linking Website Slideshow in Item, I can see the slideshow on the website but don't know how to enable the zoom feature.

image

File: erpnext/templates/generators/item.html [hunk 6]
```diff
@@ -3,122 +3,221 @@
...
+                        </div>
+                    </div>
+                </div>
+        </div>
+        {% else %}
+            {% if item_zooming %}
+                    <div class="col-md-12">
+                        <div class="col-md-3">
+                            <style>
+                                div.list_all_images img:hover {
```
Comment (netchampfaris) [inline erpnext/templates/generators/item.html]: Why inline styles tags, we can put them in .less files

File: erpnext/templates/generators/item.html [hunk 7]
```diff
@@ -3,122 +3,221 @@
...
+                                    width: 300px;
+                                    -webkit-transition: all 2s ease;
+                                    -moz-transition: all 2s ease;
+                                    -ms-transition: all 2s ease;
+                                    transition: all 2s ease;
+                                }
+                                .gallery img:hover {
+                                    transform: scale(1.5);
+                                }
+                                #selectedImage{
```
Comment (netchampfaris) [inline erpnext/templates/generators/item.html]: id selectors are bad

File: erpnext/templates/generators/item.html [hunk 8]
```diff
@@ -3,122 +3,221 @@
...
+                                    margin-left: -200px;
+                                }
+                            </style>
+                            <script>
```
Comment (netchampfaris) [inline erpnext/templates/generators/item.html]: Move to .js file?

File: erpnext/templates/generators/item.html [hunk 9]
```diff
@@ -3,122 +3,221 @@
...
+                                function changeImage(image_src){
```
Comment (netchampfaris) [inline erpnext/templates/generators/item.html]: no camelcase

File: patches.txt [hunk 10]
```diff
@@ -0,0 +1 @@
+bench.patches.v3.deprecate_old_config
```
Comment (netchampfaris) [inline patches.txt]: Remove file

**Review by netchampfaris — CHANGES_REQUESTED** [review]
Why are we using another table when we can attach multiple images to the Item form itself?
Also, there are spaces in some files, please convert to tabs.

**Comment by nabinhait** [comment]
@Solufyin Closing this, please reopen after making changes as per @netchampfaris's suggestion.

**Pull request closed without merging**